Docs: Glue Exporter and related (#6823)
treeverse/lakeFS: docs/howto/catalog_exports.md

---
title: Data Catalogs Export
description: This section explains how lakeFS can integrate with external Data Catalogs via metastore update operations. 
parent: How-To
---

# Data Catalogs Export

{% include toc_2-3.html %}

## About Data Catalogs Export

Data Catalog Export is all about integrating query engines (like Spark, AWS Athena, Presto, etc.) with lakeFS.

Data Catalogs (such as Hive Metastore or AWS Glue) store metadata for services (such as Spark, Trino and Athena). They contain metadata such as the location of the table, information about columns, partitions and much more.

With Data Catalog Exports, one can leverage the versioning capabilities of lakeFS in external data warehouses and query engines to access tables with branches and commits. 

At the end of this guide, you will be able to query lakeFS data from Athena, Trino and other catalog-dependent tools:

```sql

> use main;
> use my_branch; -- any branch
> use v101; -- or tag

Reviewer: I prefer to keep SQL keywords in uppercase.  Obviously this is not required -- but ideally we would keep some consistency.  Here I would expect "select * from users" if I said "use main".

```suggestion
USE main;
USE my_branch; -- any branch
USE v101; -- or tag
```

SELECT * FROM users 
INNER JOIN events 
ON users.id = events.user_id; -- SQL stays the same, branch or tag exist as schema
```

## How it works 

> There are several well known formats that exist today that could help export existing tables in lakeFS into a "native" object store representation
> which don't require copying the data outside of lakeFS.

Reviewer: Usually the active voice is easier to read than the passive voice.  So here I would prefer something like
```suggestion
Several well known formats exist today let you export existing tables in lakeFS into a "native" object store representation
which does not require copying the data outside of lakeFS.
```

> Since these are metadata representations, they are applied using hooks to automate the process.

Reviewer: ```suggestion
These are metadata representations and can be applied automatically through hooks.
```

### Table Decleration 

> Once a a lakeFS repository is created, the tables should be configured as a table descriptor object on the repository on the path `_lakefs_tables/TABLE.yaml`.
> Note: only tables of `type: hive` are supported and more are expected to be added. 

Reviewer: Again, the active voice would be
```suggestion
After creating a lakeFS repository, configure tables as table descriptor objects on the repository on the path `_lakefs_tables/TABLE.yaml`.
Note: the Glue exporter can currently only export tables of `type: hive`.  We expect to add more.
```

#### Hive tables

> Hive metadata server tables are essentially just a set of objects that share a prefix, with no table metadata stored on the object store.  You need to configure prefix, partitions, and schema.

Reviewer: IIRC we already define "hive tables" in our docs elsewhere.  Can we unify these definitions?  (Fine to open an issue to do this in a future PR)

Author: #6863 

You're right but the current definition is so slim 1 sentence and an example so I prefer not to unify this currently since it's really a per-case definition. 
I'll open an issue because I prefer to have a more mature and organized definition when doing that and not ad-hoc in this PR.

```yaml
name: animals
type: hive
path: path/to/animals/
partition_columns: ['year']
schema:
  type: struct
  fields:
    - name: year
      type: integer
      nullable: false
      metadata: {}
    - name: page
      type: string
      nullable: false
      metadata: {}
    - name: site
      type: string
      nullable: true
      metadata:
        comment: a comment about this column
```

Note: Useful types recognized by Hive include `integer`, `long`, `short`, `string`, `double`, `float`, `date`, and `timestamp`.
{: .note }

### Catalog Exporters 

> Exporters are code packages written in lua, each exporter is exposed as a lua function under the package namespace `lakefs/catalogexport`, they are reusable hooks to connect various types of tables to different catalogs.

Reviewer: ```suggestion
Exporters are code packages accessible through [Lua integration][LINK TO LUA INTEGRATION PLEASE].   Each exporter is exposed as a Lua function under the package namespace `lakefs/catalogexport`.  Call them from hooks to connect lakeFS tables to various catalogs.
```

Author: 👍

> [configuration]({% link reference/configuration.md %}).

Reviewer: Stray link?

Author: 👍

> Note: Check the [lua Library reference]({% link howto/hooks/lua.md %}#lua-library-reference) library for code reference, specifically everything under the prefix `lakefs/catalogexport`. 

Reviewer: I don't think we say "note" before a `{: .note }`.  I looked at 3 examples and none of them did that.

Author: 👍

{: .note }

> #### Currently supported Exporters: 

Reviewer: ```suggestion
#### Currently supported exporters
```

- Symlink Exporter: Writes metadata for the table using Hive's [SymlinkTextInputFormat](https://svn.apache.org/repos/infra/websites/production/hive/content/javadocs/r2.1.1/api/org/apache/hadoop/hive/ql/io/SymlinkTextInputFormat.html)
- AWS Glue Catalog (+ Athena) Exporter: Creates a table in Glue using Hive's format and updates the location to symlink files (reuses Symlink Exporter).
- See a step by step guide on how to integrate with [Glue Exporter]({% link integrations/glue_metastore.md %})

#### Running an Exporter  

Exporters are meant to run as [Lua hooks]({% link howto/hooks/lua.md %}).
                                                                                         

> Actions trigger can be configured with [events and branches]({% link howto/hooks/index.md %}#action-file-schema), incase custom filtering logic is required it can be achieved in the Lua script itself.

Reviewer: ```suggestion
Configure the actions trigger by using [events and branches]({% link howto/hooks/index.md %}#action-file-schema).  Of course, you can add additional custom filtering logic to the Lua script if needed.
```

Author: 👍

---
title: Data Catalogs Export
description: This section explains how lakeFS can integrate with external Data Catalogs via metastore update operations. 
parent: How-To
---

# Data Catalogs Export

{% include toc_2-3.html %}

## About Data Catalogs Export

Data Catalog Export is all about integrating external Data Warehouses (i.e AWS Athena) with lakeFS.

Data Catalogs such as Hive, Glue and other store metadata for services (such as Spark, Trino and Athena). They contain metadata such as the location of the table, information about columns, partitions and much more.

> Given the native integration between Spark and SQL, it’s most common that you’ll interact with tables in Spark environment.
> 
> While different systems support reading directly from lakefs (i.e Gateway endpoint `s3a://`) many systems only support direct access to blockstore (i.e S3 bucket). 

Reviewer 2: ```suggestion
Given the native integration between Spark and SQL, it’s most common that you’ll interact with tables in Spark environment.

While different systems support reading directly from lakeFS (using the lakeFS S3-compatible API), many systems only support direct access to blockstore (i.e S3 bucket). 
```

> While different systems support reading directly from lakefs (i.e Gateway endpoint `s3a://`) many systems only support direct access to blockstore (i.e S3 bucket). 

Author: You are right, I just deleted them, not sure what was my point.

> With Data Catalog Exports, one can leverage the versioning capabilities of lakeFS in external data warehouses and manage tables with branches and commits. 

Reviewer 2: Again, I think it's query engines - not data warehouses (I'm not an expert in the data ecosystem - so maybe @ozkatz can come up with a better term). Also, this sentence can be merged into the first sentence ("Data Catalog Export is all about...")

Author: I added "and query engines" because Snowflake for example is a data warehouse where we plan to add integration for their catalog while Athena is indeed a query engine.

> Once hooks are set up, querying lakeFS data from e.g. Athena, Trino and other catalog-dependant tools looks like this:

Reviewer 2: ```suggestion
At the end of this guide, you will be able to query lakeFS data from Athena, Trino and other catalog-dependent tools:
```
(because hooks were not mentioned yet).
